File: host/src/test/java/dk/dtu/GameCheckpointTest.java
package dk.dtu;

import dk.dtu.domain.core.*;
import dk.dtu.domain.model.Board;
import dk.dtu.domain.model.Direction;
import dk.dtu.domain.model.Robot;
import dk.dtu.domain.program.ProgramCard;
import dk.dtu.domain.rules.api.BoardAPI;
import dk.dtu.domain.rules.api.BoardApiImpl;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.concurrent.atomic.AtomicReference;

import static dk.dtu.util.BoardTestUtils.*;
import static org.junit.jupiter.api.Assertions.*;

/**
 * @author Weihao Mo
 */
class GameCheckpointTest {

    /**
     * @author Weihao Mo
     */
    @Test
    void robotWinsAfterCompletingCheckpointsInOrder() {
        Board board = initBoardWithCheckPoints(3,3);

        Robot r= new Robot(1,0,1, Direction.E);
        r.loadProgram(List.of(ProgramCard.move1()));
        List<Robot> robots = List.of(r);
        BoardAPI api = new BoardApiImpl(board, robots);
        Game game = new Game(board, api, robots);

        AtomicReference<PlayerID> observedWinner = new AtomicReference<>();
        game.addObserver(new GameObserver() {
            @Override public void onWinnerDeclared(Game g,PlayerID winner) { observedWinner.set(winner); }
            @Override public void onGameUpdate(Game g) { }
        });

        r.loadProgram(List.of(ProgramCard.move1()));
        game.startRound();

        assertEquals(2,r.getNextCheckpoint());
        assertTrue(game.getWinner().isEmpty());

        r.loadProgram(List.of(
                ProgramCard.right(),
                ProgramCard.move1(),
                ProgramCard.left(),
                ProgramCard.move1()
        ));
        game.startRound();
        assertEquals(2,r.getX());
        assertEquals(2,r.getY());
        assertEquals(3, r.getNextCheckpoint());
        assertEquals(new PlayerID(1), game.getWinner().orElse(null));
        assertEquals(new PlayerID(1), observedWinner.get());
    }

    /**
     * @author Weihao Mo
     */
    @Test
    void robotWinsAfterCompletingCheckpointsInWrongOrder() {
        Board board = initBoardWithCheckPointsInDifferentNumber(3,3);

        Robot r= new Robot(1,0,1, Direction.E);
        r.loadProgram(List.of(ProgramCard.move1()));
        List<Robot> robots = List.of(r);
        BoardAPI api = new BoardApiImpl(board, robots);
        Game game = new Game(board, api, robots);

        AtomicReference<PlayerID> observedWinner = new AtomicReference<>();
        game.addObserver(new GameObserver() {
            @Override public void onWinnerDeclared(Game g,PlayerID winner) { observedWinner.set(winner); }
            @Override public void onGameUpdate(Game g) { }
        });

        r.loadProgram(List.of(ProgramCard.move1()));
        game.startRound();

        assertEquals(1,r.getNextCheckpoint());
        assertTrue(game.getWinner().isEmpty());

        r.loadProgram(List.of(
                ProgramCard.right(),
                ProgramCard.move1(),
                ProgramCard.left(),
                ProgramCard.move1()
        ));
        game.startRound();
        assertEquals(2,r.getX());
        assertEquals(2,r.getY());
        assertEquals(2,r.getNextCheckpoint());
        assertTrue(game.getWinner().isEmpty());
        assertNull(observedWinner.get());
    }

    /**
     * @author Weihao Mo
     */
    @Test
    void robotsDoNotWinWhenSecondRobotSkipsCheckpoints() {
        Board board = initBoardWithCheckPointsInThreeDifferentNumber(3,3);

        Robot r1 = new Robot(1, 0, 1, Direction.E);
        Robot r2 = new Robot(2, 1, 2, Direction.E);
        List<Robot> robots = List.of(r1, r2);
        BoardAPI api = new BoardApiImpl(board, robots);
        Game game = new Game(board, api, robots);

        AtomicReference<PlayerID> observedWinner = new AtomicReference<>();
        game.addObserver(new GameObserver() {
            @Override public void onWinnerDeclared(Game g,PlayerID winner) { observedWinner.set(winner); }
            @Override public void onGameUpdate(Game g) { }
        });

        r1.loadProgram(List.of(ProgramCard.move1()));
        r2.loadProgram(List.of(ProgramCard.move1()));
        game.startRound();

        assertEquals(1, r1.getX());
        assertEquals(1, r1.getY());
        assertEquals(2, r1.getNextCheckpoint());
        assertEquals(1, r2.getNextCheckpoint());
        assertTrue(game.getWinner().isEmpty());
        assertNull(observedWinner.get());

        r1.loadProgram(List.of(ProgramCard.right(), ProgramCard.move1()));
        r2.loadProgram(List.of());
        game.startRound();

        assertEquals(1, r1.getX());
        assertEquals(2, r1.getY());
        assertEquals(2,r2.getX());
        assertEquals(2, r2.getY());
        assertEquals(3, r1.getNextCheckpoint());
        assertEquals(1, r2.getNextCheckpoint());
        assertTrue(game.getWinner().isEmpty());
        assertNull(observedWinner.get());
    }
}

File: engine/test/gameCheckpoint.test.ts
import { describe, it, expect } from "vitest";
import { Robot } from "../src/model/robot.js";
import { Direction } from "../src/model/direction.js";
import { ProgramCard } from "../src/program/programCard.js";
import { Game } from "../src/core/game.js";
import type { GameObserver } from "../src/core/gameObserver.js";
import { BoardApiImpl } from "../src/rules/boardApi.js";
import {
  initBoardWithCheckPoints,
  initBoardWithCheckPointsInDifferentNumber,
  initBoardWithCheckPointsInThreeDifferentNumber,
} from "./util/boardTestUtils.js";

function winnerCapturingObserver(sink: { value: number | null }): GameObserver {
  return {
    onWinnerDeclared(_g, winner) {
      sink.value = winner;
    },
    onGameUpdate() {},
  };
}

/** Ported from dk.dtu.GameCheckpointTest. */
describe("GameCheckpoint", () => {
  it("robot wins after completing checkpoints in order", () => {
    const board = initBoardWithCheckPoints(3, 3);
    const r = new Robot(1, 0, 1, Direction.E);
    r.loadProgram([ProgramCard.move1()]);
    const robots = [r];
    const api = new BoardApiImpl(board, robots);
    const game = new Game(board, api, robots);

    const observed = { value: null as number | null };
    game.addObserver(winnerCapturingObserver(observed));

    r.loadProgram([ProgramCard.move1()]);
    game.startRound();

    expect(r.getNextCheckpoint()).toBe(2);
    expect(game.getWinner()).toBeNull();

    r.loadProgram([
      ProgramCard.right(),
      ProgramCard.move1(),
      ProgramCard.left(),
      ProgramCard.move1(),
    ]);
    game.startRound();
    expect(r.getX()).toBe(2);
    expect(r.getY()).toBe(2);
    expect(r.getNextCheckpoint()).toBe(3);
    expect(game.getWinner()).toBe(1);
    expect(observed.value).toBe(1);
  });

  it("robot does not win completing checkpoints in wrong order", () => {
    const board = initBoardWithCheckPointsInDifferentNumber(3, 3);
    const r = new Robot(1, 0, 1, Direction.E);
    r.loadProgram([ProgramCard.move1()]);
    const robots = [r];
    const api = new BoardApiImpl(board, robots);
    const game = new Game(board, api, robots);

    const observed = { value: null as number | null };
    game.addObserver(winnerCapturingObserver(observed));

    r.loadProgram([ProgramCard.move1()]);
    game.startRound();

    expect(r.getNextCheckpoint()).toBe(1);
    expect(game.getWinner()).toBeNull();

    r.loadProgram([
      ProgramCard.right(),
      ProgramCard.move1(),
      ProgramCard.left(),
      ProgramCard.move1(),
    ]);
    game.startRound();
    expect(r.getX()).toBe(2);
    expect(r.getY()).toBe(2);
    expect(r.getNextCheckpoint()).toBe(2);
    expect(game.getWinner()).toBeNull();
    expect(observed.value).toBeNull();
  });

  it("robots do not win when second robot skips checkpoints", () => {
    const board = initBoardWithCheckPointsInThreeDifferentNumber(3, 3);
    const r1 = new Robot(1, 0, 1, Direction.E);
    const r2 = new Robot(2, 1, 2, Direction.E);
    const robots = [r1, r2];
    const api = new BoardApiImpl(board, robots);
    const game = new Game(board, api, robots);

    const observed = { value: null as number | null };
    game.addObserver(winnerCapturingObserver(observed));

    r1.loadProgram([ProgramCard.move1()]);
    r2.loadProgram([ProgramCard.move1()]);
    game.startRound();

    expect(r1.getX()).toBe(1);
    expect(r1.getY()).toBe(1);
    expect(r1.getNextCheckpoint()).toBe(2);
    expect(r2.getNextCheckpoint()).toBe(1);
    expect(game.getWinner()).toBeNull();
    expect(observed.value).toBeNull();

    r1.loadProgram([ProgramCard.right(), ProgramCard.move1()]);
    r2.loadProgram([]);
    game.startRound();

    expect(r1.getX()).toBe(1);
    expect(r1.getY()).toBe(2);
    expect(r2.getX()).toBe(2);
    expect(r2.getY()).toBe(2);
    expect(r1.getNextCheckpoint()).toBe(3);
    expect(r2.getNextCheckpoint()).toBe(1);
    expect(game.getWinner()).toBeNull();
    expect(observed.value).toBeNull();
  });
});
